### PR TITLE
Trajectory Optimization API

### PR DESCRIPTION
**Pull request by alberthli**
This PR adds a very generic and flexible API for trajectory optimization. As a specific instantiation, it also implements the extremely simple [predictive sampling algorithm](https://arxiv.org/pdf/2212.00541.pdf).

- defines a generic API for trajectory optimization
- defines a generic API for cost functions (and their gradients) along with the ability to pass a generic set of caches/parameters to relevant functions
    - implements a simple quadratic cost function with fixed goal state
- implements predictive sampling

### CODE REVIEW

File: ambersim/trajopt/shooting.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    """
+    d = mjx.make_data(m)
+    d = d.replace(qpos=q0, qvel=v0)  # setting the initial state.
+
+    def scan_fn(d, u):
+        """Integrates the model forward one step given the control input u."""
+        d = d.replace(ctrl=u)
+        d = step(m, d)
+        x = jnp.concatenate((d.qpos, d.qvel))  # (nq + nv,)
+        return d, x
```
Comment (alberthli) [inline ambersim/trajopt/shooting.py:43]: I'm wondering whether this is the most reasonable implementation of shooting - is `scan` performant with `mjx`'s `step` function?

Also, I'd like to double check that `step` is what I want here instead of just `forward` - I believe `forward` only computes the forward dynamics but doesn't actually resolve the equality constraints with the solver, which is in general what we need the integrator for.

File: ambersim/trajopt/shooting.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        m: The model.
+        q0: The initial generalized coordinates.
+        v0: The initial generalized velocities.
+        us: The control inputs.
+
+    Returns:
+        qs (shape=(N + 1, nq)): The generalized coordinates.
+        vs (shape=(N + 1, nv)): The generalized velocities.
+    """
+    d = mjx.make_data(m)
```
Comment (alberthli) [inline ambersim/trajopt/shooting.py:34]: I make a new `Data` for every call to the `shoot` function (and in a `vmap`, the fields of `d` should become `vmap`ped as well, I believe). Is it really slow to do this? If so, how else should I be managing `mjx.Data`?

File: ambersim/trajopt/shooting.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        # TODO(ahl): check whether joints with no limits have reasonable defaults for m.actuator_ctrlrange
+        limits = m.actuator_ctrlrange
+        clip_fn = partial(jnp.clip, a_min=limits[:, 0], a_max=limits[:, 1])  # clipping function with limits already set
+        us_samples = vmap(vmap(clip_fn))(_us_samples)  # apply limits only to the last dim, need a nested vmap
+        # limited = m.actuator_ctrllimited[:, None]  # (nu, 1) whether each actuator has limited control authority
+        # default_limits = jnp.array([[-1000.0, 1000.0]] * m.nu)  # (nu, 2) default limits for each actuator
+        # limits = jnp.where(limited, m.actuator_ctrlrange, default_limits)  # (nu, 2)
+
+        # predict many samples, evaluate them, and return the best trajectory tuple
+        qs_samples, vs_samples = vmap(shoot, in_axes=(None, None, None, 0))(m, q0, v0, us_samples)
```
Comment (alberthli) [inline ambersim/trajopt/shooting.py]: Just from casually commenting stuff out, i've 100% confirmed that it's this `vmap` over `shoot` call which takes up nearly 100% of the runtime (not the stuff before or below it).

File: ambersim/trajopt/shooting.py (hunk 2 above)
Comment (alberthli) [reply ambersim/trajopt/shooting.py:34]: Tested: doesn't make a difference. The main difference is in the `step` call.

File: ambersim/trajopt/shooting.py (hunk 1 above)
Comment (alberthli) [reply ambersim/trajopt/shooting.py:43]: Tested: `scan` runs just as well as a native `python` `for` loop. `step` is also definitely the right function.

File: examples/trajopt/ex_predictive_sampling.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
+import jax
```
Comment (alberthli) [inline examples/trajopt/ex_predictive_sampling.py]: You can defer reviewing this file to #47.

Comment (vincekurtz) [reply examples/trajopt/ex_predictive_sampling.py]: In that case let's drop this file from this PR.

However, I do think it'd be good to have a simple example that verifies that predictive sampling is doing the right thing (even if it's inefficient at this point). The pendulum example should work with only a few samples, and should be pretty easy to spin up.

File: ambersim/trajopt/base.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+        The shapes of the outputs include (?) because we may choose to return non-zero-order-hold parameterizations of
+        the optimized trajectories (for example, we could choose to return a cubic spline parameterization of the
+        control inputs over the trajectory as is done in the gradient-based methods of MJPC).
+
+        Args:
+            params: The parameters of the trajectory optimizer.
+
+        Returns:
+            qs_star (shape=(N + 1, nq) or (?)): The optimized trajectory.
```
Comment (vincekurtz) [inline ambersim/trajopt/base.py]: Do we want the user to think about `q, v, u` or do we want the user to think about `x, u`? IMO `x, u` makes more sense from a user perspective. Separating `q` and `v` makes sense for IDTO but we might want to abstract those details away from the user. 

Overall I'm fine either way as long as it's consistent.

File: ambersim/trajopt/cost.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        """Computes a batched quadratic form for a single instance of A.
+
+        Args:
+            bs (shape=(..., n)): The batch of vectors.
+            A (shape=(n, n)): The matrix.
+
+        Returns:
+            val (shape=(...,)): The batch of quadratic forms.
+        """
+        return jnp.einsum("...i,ij,...j->...", bs, A, bs)
```
Comment (vincekurtz) [inline ambersim/trajopt/cost.py:47]: btw, this is cool! I had no idea this existed

**Review by vincekurtz — CHANGES_REQUESTED** [review]
Overall API looks good, just missing a few tests. 

I would add tests for:
- Computing the cost for quadratic with static goal (vs for loop)
- Computing the gradient and Hessian for the quadratic with static goal (vs autodiff)
- A smoke test for creating a `VanillaPredictiveSampler` and calling `optimize` (maybe not necessary depending on below)
- Some sanity check that predictive sampling is doing the right thing. That could be a pendulum swingup example (should work even if the implementation isn't blazing fast yet) or a little unit test that shows the cost generally going down, or something else.

**Comment by alberthli** [comment]
Agreed on the first two requested tests - those should be trivial to spin up.

Sanity check of choice: randomly sample a batch of initial policies and also shoot them forward. pass those guesses to vanilla predictive sampling, which returns new trajectories that should be no worse than the guesses. verify this property in a test.

TODO list:
- [x] refactor `q` and `v` to `x` everywhere
- [x] test cost function and its derivatives (thanks for suggesting this, literally all my implementations were wrong)
- [x] smoke test for `VanillaPredictiveSampler` + `optimize` + `jit`
- [x] sanity check for `VanillaPredictiveSampler`
- [x] drop example file from PR after all other tests are implemented

File: ambersim/trajopt/cost.py (hunk 6 above)
Comment (alberthli) [reply ambersim/trajopt/cost.py:47]: I actually use it moderately frequently, but for this specific instance, you can thank Github copilot!

File: ambersim/trajopt/base.py (hunk 5 above)
Comment (alberthli) [reply ambersim/trajopt/base.py]: It's a good question, there's definitely more annoying bookkeeping for the end user if we spit out `q` and `v` instead of `x` (and it makes managing gradients more annoying for devs). `x` and `u` is certainly more general, so I'm OK refactoring over to it and managing the interface with `mjx` on the dev side.

Comment (alberthli) [reply ambersim/trajopt/base.py]: Refactored.

File: examples/trajopt/ex_predictive_sampling.py (hunk 4 above)
Comment (alberthli) [reply examples/trajopt/ex_predictive_sampling.py]: Dropped.

File: tests/trajopt/test_predictive_sampler.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    num_steps = 10
+    us_guess = jnp.zeros((num_steps, model.nu))
+    params = VanillaPredictiveSamplerParams(key=key, x0=x0, us_guess=us_guess)
+
+    # sampling the best sequence of qs, vs, and us
+    optimize_fn = jit(ps.optimize)
+    assert optimize_fn(params)
+
+
+def test_VPS_cost_decrease():
```
Comment (alberthli) [inline tests/trajopt/test_predictive_sampler.py]: self nit: i could probably make these tests faster by not initializing the same model + data for multiple tests.

**Review by vincekurtz — APPROVED** [review]
LGTM
